File: Medium/17. Letter Combinations of a Phone Number/letterCombinations.cpp
/*
    time complexity: O(4^n)
    space complexity: O(4^n + n)

*/
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if(digits.empty()){
            return {};
        }
        vector<vector<char>> d(10);
        d[0] = {' '};
        d[1] = {};
        d[2] = {'a','b','c'};
        d[3] = {'d','e','f'};
        d[4] = {'g','h','i'};
        d[5] = {'j','k','l'};
        d[6] = {'m','n','o'};
        d[7] = {'p','q','r','s'};
        d[8] = {'t','u','v'};
        d[9] = {'w','x','y','z'};
        string cur = "";
        vector<string> ans;
        dfs(digits, ans, cur, d, 0);
        return ans;
    }
private:
    void dfs(string &digits, vector<string> &ans, string &cur, vector<vector<char>> &d, int l){
        if(l == digits.size()){
            ans.push_back(cur);
            return;
        }
        for(char c : d[digits[l] - '0']){
            cur.push_back(c);
            dfs(digits, ans, cur, d, l+1);
            cur.pop_back();
        }
    }
};

File: Medium/17. Letter Combinations of a Phone Number/letterCombinations.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if(digits.length() == 0){
            return new ArrayList<>();
        }
        String [] dict = {"", "", "abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        List<String> ans = new ArrayList<>();
        dfs(digits.toCharArray(), dict, ans, "");
        return ans;
    }
    private void dfs(char [] digits, String[] m, List<String> ans, String s){
        if(s.length() == digits.length){
            ans.add(s);
            return;
        }
        int i = s.length();
        int d = digits[i] - '0';
        for(char letter : m[d].toCharArray()){
            dfs(digits, m, ans, s + Character.toString(letter));
        }
    }
}